src/libcharon/plugins/kernel_super/kernel_super_ipsec.c

```diff
@@ -305,6 +305,45 @@ static void write_tunnel_oper_items(private_kernel_super_ipsec_t* this,
 	LOG(DBG_KNL, SL_LOG_ERR, "Failed to write tunnel");
 }
 
+static bool get_tunnel_lifetime_intervals(ike_sa_t *ike_sa, struct terminator_tunnel_lifetime_intervals *intervals) {
+	peer_cfg_t *peer_cfg = ike_sa->get_peer_cfg(ike_sa);
+	if(! peer_cfg) {
+		LOG(DBG_KNL, SL_LOG_ERR, "Failed to get peer_cfg");
+		return false;
+	}
+
+	child_cfg_t *child_cfg = NULL;
+	enumerator_t *enumerator = peer_cfg->create_child_cfg_enumerator(peer_cfg);
+	enumerator->enumerate(enumerator, &child_cfg);
+	if(! child_cfg) {
+		LOG(DBG_KNL, SL_LOG_ERR, "Failed to get child_cfg");
+		return false;
+	}
+	enumerator->destroy(enumerator);
+
+	intervals->ike_rekey_interval = peer_cfg->get_rekey_time(peer_cfg, FALSE);
+	intervals->ike_reauth_interval = peer_cfg->get_reauth_time(peer_cfg, FALSE);
```

Comment:
I don't like the fact that you initialize `intervals` into a "zombie" state by setting its fields with incorrect values which you later fixup.
Given that the final initialization is set only a few lines afterwards makes it less readable to understand when exactly the fields are correct, and more error-prone since future code may be added here which uses say `intervals->ike_reauth_interval` (since this is the place where it is set), but the person writing it will not know / forget that this value is actually incorrect here, and only a few lines down it is corrected.

As a general rule, opt to make variables be initialized with correct values, and if need to then perform any needed calculation before the initialization.
In this case, first save all the values from the `peer_cfg` and `child_cfg` into separate variables:
``` c
const uint32_t ike_reauth_time = peer_cfg->get_reauth_time(peer_cfg, FALSE);
const uint32_t ike_rekey_time = peer_cfg->get_rekey_time(peer_cfg, FALSE);
const uint32_t rekey_margin = peer_cfg->get_over_time(peer_cfg);
const uint32_t ipsec_rekey_time = child_cfg_lifetime->time.rekey;
```
and only then initialize the values in `intervals`:
``` c
intervals->ike_reauth_interval = (0 == ike_reauth_time) ? 0 : ike_reauth_time + rekey_margin;
intervals->ike_rekey_interval = (0 == ike_rekey_time) ? 0 : ike_rekey_interval + rekey_margin;
intervals->ipsec_rekey_interval = (0 == ipsec_rekey_time) ? 0 : ipsec_rekey_time + rekey_margin;
intervals->rekey_margin = rekey_margin;
```